src/main/java/tweeting/services/Twitter4JService.java

```diff
@@ -22,53 +22,76 @@
 
     private static final Logger logger = LoggerFactory.getLogger(Twitter4JService.class);
 
+    private TimelineCache homeTimelineCache;
+
     public Twitter4JService(Twitter api) {
         this.api = api;
+        homeTimelineCache = new TimelineCache();
     }
 
     @Override
-    public Optional<Tweet> postTweet(String message)
+    public Tweet postTweet(String message)
             throws TwitterServiceResponseException, TwitterServiceCallException {
         if (StringUtils.isBlank(message)) {
             throw new TwitterServiceCallException(MISSING_TWEET_MESSAGE);
         }
-        if (message.length() > MAX_TWEET_LENGTH) {
+        else if (message.length() > MAX_TWEET_LENGTH) {
             throw new TwitterServiceCallException(TOO_LONG_TWEET_MESSAGE);
         }
         try {
             return Optional.ofNullable(api.updateStatus(message))
                     .map(status -> {
                         logger.info("Successfully posted '{}' to Twitter.", message);
-                        return constructTweet(status);
-                    });
+                        final Tweet tweet = constructTweet(status);
+                        homeTimelineCache.invalidate();
+                        return tweet;
+                    }).get();
         } catch (TwitterException te) {
             throw createServerException(te);
         }
     }
 
+
     @Override
-    public Optional<List<Tweet>> getHomeTimeline() throws TwitterServiceResponseException {
+    public List<Tweet> getHomeTimeline() throws TwitterServiceResponseException {
         try {
-            return Optional.ofNullable(api.getHomeTimeline())
-                    .map(statuses -> {
-                        logger.info("Successfully retrieved home timeline.");
-                        return constructTweetList(statuses);
-                    });
+            final Optional<List<Tweet>> cachedTweets = homeTimelineCache.getCachedTimeline("");
+            if (cachedTweets.isPresent()) {
+                logger.info("Successfully retrieved home timeline from cache.");
+                return cachedTweets.get();
+            }
+            final List<Tweet> tweets = api.getHomeTimeline()
+                    .stream()
+                    .map(this::constructTweet)
+                    .filter(Objects::nonNull)
+                    .collect(Collectors.toList());
+            homeTimelineCache.cacheTimeline("", tweets);
+            logger.info("Successfully retrieved home timeline from Twitter.");
+            return tweets;
         } catch (TwitterException te) {
             throw createServerException(te);
         }
     }
 
+    //logger.info("Successfully retrieved home timeline filtered by \'" + keyword + "\' from cache.");
     @Override
-    public Optional<List<Tweet>> getFilteredTimeline(String keyword)
+    public List<Tweet> getFilteredTimeline(String keyword)
```

Comment:
why the removal of the optional here when you are wrapping the result in an optional in the resource class